src/tyr/serializers.cc

```diff
@@ -173,12 +172,31 @@ waypoint(const valhalla::Location& location, bool is_tracepoint, bool is_optimiz
   if (is_optimized) {
     int trips_index = 0; // TODO
     waypoint->emplace("trips_index", static_cast<uint64_t>(trips_index));
-    waypoint->emplace("waypoint_index", static_cast<uint64_t>(location.shape_index()));
+    waypoint->emplace("waypoint_index", static_cast<uint64_t>(location.waypoint_index()));
   }
 
   return waypoint;
 }
 
+/*
+ * This function serializes the via_waypoints object on the leg.
+ */
+valhalla::baldr::json::MapPtr serialize_via_waypoint(const valhalla::Location& origin,
+                                                     const valhalla::Location& location,
+                                                     const uint64_t geometry_idx) {
+
+  // Create a via waypoint to add to the array
+  auto via_waypoint = json::map({});
+  // Add distance in meters from the input location to the silent waypoint
+  via_waypoint->emplace("distance_from_leg_start",
+                        json::fixed_t{to_ll(origin.ll()).Distance(to_ll(location.path_edges(0).ll())),
```

Comment:
`PointLL::Distance` measerus the straight line distance which is not what we want here. We want the distance along the route-line.